### PR TITLE
refactor with invalidatedRevision

### PR DESCRIPTION
**Pull request by dai-shi**
This might not change much for now.
Hope this is better with real uMS and in CM.
At least the code is cleaner. 😁

### CODE REVIEW

File: src/core/vanilla.ts [hunk 1]
```diff
@@ -212,10 +216,21 @@ const setAtomReadPromise = <Value>(
     atom,
     dependencies
   ) as AtomState<Value>
+  delete atomState.i
   atomState.p = promise
   return new Map(wip).set(atom, atomState)
 }
 
+const setAtomInvalidatetd = <Value>(
```
Comment (Thisen) [inline src/core/vanilla.ts:224]:
```suggestion
const setAtomInvalidated = <Value>(
```

File: src/core/vanilla.ts [hunk 2]
```diff
@@ -212,10 +216,21 @@ const setAtomReadPromise = <Value>(
...
+  state: State,
+  wip: WorkInProgress,
+  atom: Atom<Value>
+): WorkInProgress => {
+  const atomState = wipAtomState(state, wip, atom) as AtomState<Value>
+  atomState.i = atomState.r
+  return new Map(wip).set(atom, atomState)
```
Comment (Thisen) [inline src/core/vanilla.ts:231]: Is `concatMap` better here?

File: src/core/vanilla.ts (hunk 1 above)
Comment (dai-shi) [reply src/core/vanilla.ts:224]: 😓

File: src/core/vanilla.ts (hunk 2 above)
Comment (dai-shi) [reply src/core/vanilla.ts:231]: It's just one set, so this should be good.

**Pull request merged**